### PR TITLE
fix(ci): reconcile programme and Wave 0 truth

### LINKED ISSUE #2515: ci: reconcile stale Wave 0 gate and programme state
## Problem

The durable split/CI surfaces have drifted from current repository and GitHub state.

Re-measured against `origin/main` at `974336892ce7721dcf32e8ad5d2160ea4ec11408` on 2026-08-18:

- `AGENTS.md` still names #2388 as the active programme ledger, but #2388 has been closed since 2026-08-15. No newer open programme/wave/refactor ledger was found by title search.
- `docs/contributing/WAVE0-GATES.md` still documents the removed `WAVE0_SEMVER_BASELINE_SHA` and fixed-SHA reset process. The workflow now resolves the newest release tag dynamically and protects that behavior with `scripts/ci/test-semver-gate.sh`.
- The same checklist recommends three Wave 0 jobs as required checks, while live `main` branch protection requires `CI`, `host-capability-check`, and `lane-check/proof`.
- The warn-only unsafe preview in `.github/workflows/split-wave0-gates.yml` still treats `monitor.rs` as the only intended boundary and proposes promoting that assumption. `monitor.rs` still exists, but current code has several intentional unsafe boundaries; the single-boundary message is unsupported and noisy.
- `scripts/ci/review-split-wave.sh` inventories `base...HEAD` and staged changes but omits unstaged/untracked state. Its current Assay Sim example still resolves and is not part of the defect.
- `docs/contributing/REFACTOR-WAVE-STATUS.md` calls Wave71 active. Its closure gate exists, but it failed with 49 handwritten Rust files at or above 600 LOC on `0e934b718f44db3554d9e5cf79d73f5d7180d351`. That supports “incomplete”, not “closed”; without current execution evidence, it should not say “Active”.

The Split Wave 0 workflow itself is not dead. It still runs on each PR, generally skips heavy jobs for irrelevant changes, and retains unique feature-matrix and semver coverage.

## Proposed scope

1. Reset the canonical ledger line in `AGENTS.md` to “no active programme”, unless a newer ledger is explicitly selected.
2. Update `WAVE0-GATES.md` to describe the dynamic latest-release baseline and distinguish actual required contexts from historical recommendations.
3. Remove the generic unsafe preview and its unsupported single-boundary promotion message. Add tested unsafe allowlists only in explicitly scoped split/boundary gates.
4. Keep `review-split-wave.sh`, but fail on unstaged/untracked state or include it in scope inventory. Preserve examples that resolve on the current tree.
5. Mark Wave71 dormant/incomplete (or provide current active execution evidence). Retain the LOC script as a measurement/selection tool and rerun it on the implementation head.

## Acceptance criteria

- Public programme-ledger state points to an open ledger or plainly says none is active.
- Wave 0 semver documentation matches the workflow and its contract test.
- Required-check documentation matches live protection or is clearly identified as non-current advice.
- The unsupported generic single-boundary unsafe preview/message is gone.
- The generic split reviewer cannot silently ignore dirty worktree files; existing valid examples remain valid.
- Wave71 status does not claim active execution without a current authoritative ledger.
- Focused script/contract tests, `git diff --check`, and public-surface string review pass on the final SHA.

## Non-goals

- Removing the feature-matrix or semver gates.
- Declaring the repository below the 600-LOC threshold.
- Introducing a repository-wide “unsafe only in one file” rule.
- Opening a new programme ledger solely for this standalone maintenance issue.
- Replacing the valid Assay Sim example.

### PR DESCRIPTION
**Pull request by Rul1an**
## Summary

- reset the public programme ledger to an explicit no-active-programme state
- align Wave 0 docs with the dynamic release baseline and canonical required-context source
- remove the unsupported generic single-unsafe-boundary preview
- inventory committed, staged, unstaged, and untracked split-wave changes through one shared function
- mark Wave71 dormant/incomplete without claiming the hotspot threshold is complete
- add focused mutation-sensitive contracts and ensure their documentation inputs trigger CI

## Why

The durable CI/refactor surfaces had drifted from repository and live branch-protection state. This keeps the useful Wave 0 gates while removing stale claims and making dirty worktree scope visible.

## Contract boundaries

- no repository-wide single-unsafe-boundary rule is introduced
- the ledger guard supports explicit inactive and structurally active states
- the offline ledger guard verifies link form and visible/URL issue-ID parity, not live GitHub OPEN/CLOSED state
- required contexts remain canonical in `CI-CONTRACT.md` and `.github/rulesets/main-required-ci-contexts.json`; Wave 0 docs point there rather than duplicating names
- current hotspot measurement remains above the target, so Wave71 is dormant/incomplete, not closed

## Verification

Exact head: `83138cbc34a5274b4539170ff43ae0d06ef4f457`

- `bash scripts/ci/test-ci-programme-truth.sh`
- `bash scripts/ci/test-review-split-wave.sh`
- `python3 scripts/ci/check-required-contexts.py`
- `python3 scripts/ci/check-required-contexts.py --self-test`
- live required-context reconciliation through `--live-response`
- `shellcheck -x scripts/ci/review-split-wave.sh scripts/ci/test-ci-programme-truth.sh scripts/ci/test-review-split-wave.sh`
- `actionlint .github/workflows/kernel-matrix.yml .github/workflows/split-wave0-gates.yml`
- `pre-commit validate-config`
- `cargo fmt --all -- --check`
- `git diff --check 974336892ce7721dcf32e8ad5d2160ea4ec11408..83138cbc34a5274b4539170ff43ae0d06ef4f457`
- full pre-push hook, including workspace clippy and CI programme-truth contract

All passed.

## Review quorum

Independent non-building Codex subagent reviewed exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`: **READY**, no actionable findings. It independently exercised active/inactive ledger transitions, mismatch/duplicate/contradiction mutations, benign prose, dirty-tree inventory, trigger removals, cleanup fault paths, exact hook-command execution, all five forbidden-override tuple deletions, and shared caller parsing. All three findings from the prior head were closed with targeted RED mutations.

Closes #2515

## Summary by CodeRabbit

* **CI & Quality**
  * CI checks now run when programme-truth documentation, workflows, rulesets, or related configuration changes.
  * Added fail-closed validation for file sizes, inventory limits, overrides, workflow triggers, and review-split behavior.
  * Review checks now include committed, staged, unstaged, and untracked files.
  * Programme-truth checks run during both pre-commit and pre-push.

* **Documentation**
  * Updated programme status, Wave 0 gate guidance, and refactor-wave documentation.
  * Removed the unsafe boundary preview requirement and related quality-gate reporting.

### CODE REVIEW

**Comment by Rul1an** [comment]
## Independent exact-head review record

- **Head:** `a85325d8e8e22b32a834f950e7317ae3071bdcfc`
- **Base:** `974336892ce7721dcf32e8ad5d2160ea4ec11408`
- **Reviewer:** independent non-building Codex subagent; authored neither the change nor issue/spec #2515
- **Verdict:** **READY**
- **Findings:** none blocking or actionable

Independent mutations verified:

- linked plus unlinked same issue-ID: RED
- active and inactive visible/URL issue-ID mismatches: RED
- duplicate ledger bullet: RED
- plain/list/blockquote/canonical-link active or inactive declarations outside the canonical bullet: RED
- stale required context outside the canonical section: RED
- benign ledger and required-check prose: GREEN
- full contract under a canonical active ledger: GREEN
- each truth-only kernel trigger removal: RED
- normal, child-failure, and allocation-failure cleanup paths: no residue

Focused contracts, required-context canonical/self/live checks, shellcheck, actionlint, pre-commit validation, `cargo fmt --all -- --check`, and `git diff --check` all passed. Worktree was clean. This review counts for the repository's one non-building-agent quorum on this exact SHA; any push invalidates it.

**Comment by Rul1an** [comment]
CodeRabbit posted three actionable exact-head findings after CI completed. They are technically valid: active-ledger canonical wording is under-validated, and the new PR-controlled inventory/document inputs need fail-closed pre-materialization ceilings per `AGENTS.md`. The PR is back in draft while Cursor fixes them locally. The existing READY review is invalidated by the next push; a new exact-head independent review will be recorded before returning to ready.

**Comment by Rul1an** [comment]
## Independent exact-head review

**Verdict: READY** at `83138cbc34a5274b4539170ff43ae0d06ef4f457` (base and merge-base `974336892ce7721dcf32e8ad5d2160ea4ec11408`).

No actionable findings.

The non-building reviewer independently closed all three findings from the prior head:

- Replacing either hook suite invocation from `bash` to `echo` is targeted RED with `ci-programme-truth hook entry mismatch`.
- Removing each of the five canonical forbidden-override tuple entries fails before self-invocation in 0.12-0.13s with `forbidden override set mismatch`; no timeout or recursion.
- Count and drop share `reject_overrides_caller_marks()`; source inspection found one normalizer and one independent five-name test table. Independent marks/count/drop mutations are RED.

Deleting the exact `reject-overrides` invocation independently from either suite is targeted RED for all five variables in 0.03-0.04s. Empty-valued direct environment probes, ROOT empty/nonexistent cases, combined override cases, ledger mutations, unsafe-preview mutations, NUL/FIFO/symlink/special-file and size-ceiling cases all remained covered.

Independent verification included both focused suites under Bash 3.2, Linux-container execution and invalid-byte filename fidelity, `shellcheck`, `actionlint`, required-context checks and self-test, `pre-commit validate-config`, the focused pre-push hook, `cargo fmt --all -- --check`, and `git diff --check`.

Reviewer declaration: the reviewer authored or edited neither this PR nor issue #2515 or a governing specification; all adversarial mutations ran in disposable copies. The reviewer did not modify a worktree, commit, push, post to GitHub, or dispatch workflows.

File: scripts/ci/review-split-wave.sh [hunk 1]
```diff
@@ -25,15 +25,22 @@ if ! git rev-parse --verify "${base_ref}" >/dev/null 2>&1; then
...
+    git diff --name-only
+    git ls-files --others --exclude-standard
+  } | sed '/^$/d' | sort -u
+}
+
+changed_files="$(split_wave_changed_files "${base_ref}")"
 
 if [[ -z "${changed_files}" ]]; then
-  echo "FAIL: no tracked changes found against ${base_ref}" >&2
+  echo "FAIL: no changes found against ${base_ref}" >&2
```
[the remark this answers is not in the document]
Comment (Rul1an) [reply scripts/ci/review-split-wave.sh:55]: Fixed on exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`. `split_wave_changed_files` streams all four Git inventories directly through the canonical `resource_ceilings.py inventory` helper before command substitution; that helper enforces caller-nonraisable path-count and byte ceilings before sorting/materialization. The test suite reads source/tracked inputs through bounded regular-file/stream helpers and independently turns RED for path-count, byte-count, oversized source, raised/invalid caps, FIFO, symlink and special-file cases. The independent exact-head reviewer re-ran these mutations.

File: scripts/ci/test-ci-programme-truth.sh [hunk 2]
```diff
@@ -0,0 +1,584 @@
...
+bad() { echo "FAIL  $1"; FAILURES=$((FAILURES + 1)); }
+
+LEDGER_PREFIX='- The public execution ledger'
+
+assert_extracted_block() {
+  local kind="$1" expected="$2" text="$3"
+  PROGRAMME_TRUTH_KIND="$kind" \
+  PROGRAMME_TRUTH_EXPECTED="$expected" \
+  PROGRAMME_TRUTH_DOC="$text" \
+  PROGRAMME_TRUTH_LEDGER_PREFIX="$LEDGER_PREFIX" python3 - <<'PY'
```
[the remark this answers is not in the document]
Comment (Rul1an) [reply scripts/ci/test-ci-programme-truth.sh]: Fixed on exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`. Full document content is no longer transported through environment variables: the Python validators receive file paths and use `read_bounded_file`, which opens one descriptor, rejects non-regular inputs, and reads at most the canonical ceiling plus one byte before any consumer sees data. Synthetic fixtures are written as bounded files, and oversize/FIFO/symlink/special-file mutations fail with controlled diagnostics.

File: scripts/ci/test-ci-programme-truth.sh [hunk 3]
```diff
@@ -0,0 +1,584 @@
...
+    if inactive_count > 1:
+        raise SystemExit("ledger bullet has multiple inactive declarations")
+    if inactive_count == 1:
+        if "active programme ledger" in bullet_c:
+            raise SystemExit("contradictory ledger claims in the ledger bullet")
+        return
+    if len(links) != 1:
+        raise SystemExit(
+            "ledger bullet is neither a valid inactive nor a valid active state"
+        )
```
[the remark this answers is not in the document]
Comment (Rul1an) [reply scripts/ci/test-ci-programme-truth.sh:293]: Fixed on exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`. The active branch now requires the collapsed bullet to start with the exact canonical declaration and requires the remaining payload to be exactly one matching `issue #N` link. Retired/historical/noncanonical one-link wording is rejected. The focused suite includes independent noncanonical-active, mismatch, duplicate, contradiction and outside-declaration mutations; the exact-head reviewer re-ran the ledger mutations.

File: scripts/ci/lib/resource_ceilings.py [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+
+def drop_reject_overrides_caller(src: str, dst: str) -> None:
+    text = read_bounded_file(src).decode("utf-8")
+    marks = reject_overrides_caller_marks(text)
+    count = sum(1 for _line, is_caller in marks if is_caller)
+    if count != 1:
+        raise SystemExit(f"reject-overrides caller count is {count}, want 1")
+    out = "".join(line for line, is_caller in marks if not is_caller).encode("utf-8")
+    require_bounded_bytes(out, dst)
+    with open(dst, "wb") as fh:
```
[the remark this answers is not in the document]
Comment (Rul1an) [reply scripts/ci/lib/resource_ceilings.py:70]: False positive on exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`. `drop-reject-caller` is test-only mutation I/O. Its sole repository callsite writes `$TRUTH_TMP/no-reject-caller-$(basename "$suite")` beneath a private `mktemp -d` (observed mode 0700) for two fixed suite paths. The helper runs with the same OS identity and no additional authority; a caller able to choose argv can already write any same-user-writable path. No service, setuid, auth, secret, or cross-user boundary is crossed. Adding path canonicalization/O_EXCL would silence taint but not reduce a reachable capability.

File: scripts/ci/lib/resource_ceilings.py [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+    path: str, label: str | None = None, limit: int = MAX_DOC_BYTES
+) -> bytes:
+    """Read at most ``limit`` bytes from a regular file via one descriptor.
+
+    Non-regular inputs fail closed. At most ``limit + 1`` bytes are read
+    from the opened fd so an oversize file never reaches a consumer.
+    """
+    label = label or path
+    try:
+        fd = os.open(path, _readonly_open_flags())
```
[the remark this answers is not in the document]
Comment (Rul1an) [reply scripts/ci/lib/resource_ceilings.py:100]: False positive on exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457`. `read_bounded_file` intentionally accepts a local caller-supplied path and opens it with the same user privileges. Repository callsites supply fixed repo paths, `BASH_SOURCE`, hermetic copies, private `mktemp` fixtures, or explicit hostile-path test fixtures; production `review-split-wave` uses only the stdin inventory command. The reader already fails closed with `O_NOFOLLOW`/`O_NONBLOCK`, one-descriptor `fstat` regular-file validation, and a limit+1 byte ceiling. There is no privilege/authorization boundary for path sanitization to protect.

**Comment by Rul1an** [comment]
## Extra-evidence sidecar (not quorum)

Exact head `83138cbc34a5274b4539170ff43ae0d06ef4f457` (base `974336892ce7721dcf32e8ad5d2160ea4ec11408`). Fresh GitHub tarball extract only; no lasting clone, no worktree write, no product edit.

**Scope.** Ten CI/docs/script files vs that base; no product crates. CodeQL path-injection alerts on `scripts/ci/lib/resource_ceilings.py` are **not** dispositioned here (they did not break this run). `shellcheck` and `actionlint` were **MISSING** on this box and were not run.

**Clean suites**

- `bash scripts/ci/test-ci-programme-truth.sh` on the gitless extract: **PASS**, exit 0, 11.5s.
- `bash scripts/ci/test-review-split-wave.sh` on the gitless extract: **FAIL** only on usage examples (`fatal: not a git repository` / `git ls-files failed`). Same suite on a disposable copy with throwaway `git init` + one extract commit: **PASS**, exit 0, 1.2s.
- `python3 -m py_compile scripts/ci/lib/resource_ceilings.py`: exit 0.

**Hook-command neutralization (disposable copies).** `.pre-commit-config.yaml` `id: ci-programme-truth` expected entry is `bash -c 'bash scripts/ci/test-ci-programme-truth.sh && bash scripts/ci/test-review-split-wave.sh'`. Replacing either inner `bash` with `echo` is targeted **RED** with `ci-programme-truth hook entry mismatch` (isolated assert 0.02s; full suite also RED). Neither neutralization was GREEN.

Measured, not a verdict: the programme-truth `bash`→`echo` path is an `expect_ok` check, so that full suite continued into later self-invoking probes after the RED; the review-split `bash`→`echo` path aborted earlier under `set -e`.

**Five forbidden-override tuple deletions (disposable copies, one name each).** Canonical set: `PROGRAMME_TRUTH_ROOT`, `PROGRAMME_TRUTH_AGENTS`, `PROGRAMME_TRUTH_SELFHOST`, `PROGRAMME_TRUTH_CEILING_CHILD`, `PROGRAMME_TRUTH_PREVIEW_MUTANT`. Deleting any one name from `FORBIDDEN_PROGRAMME_OVERRIDES` fail-closes **before** self-invocation with `forbidden override set mismatch` in 0.069–0.076s. Exit 1, no timeout, no recursion, no `PROGRAMME_TRUTH_ROOT` empty-override probe. None GREEN. The clean suite’s per-var table is independently RED for each name (not a combined-five-only assert).

**Non-claims.** Extra evidence only; not the independent exact-head review and not quorum. No branch, commit, push, dispatch, ready, merge, or release. No cargo/workspace/pre-push. Tarball is not a git repo; the review-split re-run used a throwaway git copy only. Scratch `/tmp/ruley-2523-83138cbc` deleted (~369M freed).

**Comment by Rul1an** [comment]
## Machine-readable review binding

Head: `83138cbc34a5274b4539170ff43ae0d06ef4f457`

VERDICT: READY

Reviewer: independent non-building Codex subagent. No actionable findings; prior findings are closed, all current-head review threads are resolved, and CodeQL alerts #823/#824 were independently classified and narrowly dismissed with recorded same-user/no-privilege-boundary rationale. Full evidence and independence declaration: https://github.com/Rul1an/assay/pull/2523#issuecomment-5338048003
